### ISSUE TITLE
Contact avatar pictures broken after attempting to compress pictures in the "avatar" folder

### ISSUE THREAD

**Issue by csolisr** [friendica/friendica#14047]
- [X] I have searched open and closed issues for duplicates

### Bug Description

I recently tried to save some space in my server by compressing my avatar images (I use the avatar cache files). However, now I have many broken contact pictures. I tried restoring them from an old backup to no avail. I suspect that the checksum of the files varied and that broke the bindings between pictures and avatars somehow. Is there a way to force all contacts to have their pictures re-fetched?

### Steps to Reproduce

1.  Install Friendica, and enable the avatar cache in the settings. This should create an `avatar` folder, typically in `/var/www/friendica/avatar`.
2. Make a backup of your `avatar` folder (optional).
3.  Run a script to compress the contents of the `avatar` folder with `imagemagick` for example.
4.  Return to the Friendica installation. Notice that the pictures for several users are now inaccessible, returning an error 404 "Not Found" upon inspection.
5. If you had a backup of your pictures, restore it. Notice that the pictures are still inaccessible anyways.

Actual Result:

See above. The pictures for several users are inaccessible, and in several cases they might not be restored.

Expected Result:

Pictures for contacts should be available in some form - either being re-fetched in case that the backend does not work correctly, or else using the database as a fallback.

### Screenshots

![imagen](https://github.com/friendica/friendica/assets/1644021/3fbec1e0-19b0-424c-bc54-8959925d441e)

### Platform Info

Friendica Version: 'Yellow Archangel' 2024.06-dev - 1559

Friendica Source: Compiled from Git

PHP version: 8.2

SQL version: 15.1

**Comment by MrPetovan** [comment]
Are the 404 URLs pointing to avatar pictures direct link to the files in the `avatar` folder?

**Comment by csolisr** [comment]
> Are the 404 URLs pointing to avatar pictures direct link to the files in the `avatar` folder?

Yes, they follow the pattern "/avatar/cd/7d/ad9/68/b5/9c72/91a94b2ff7123f7e-80.png?ts=1711319692" for example.

**Comment by MrPetovan** [comment]
And they all point at non-existent files?

**Comment by csolisr** [comment]
Correct, the corresponding files do not exist in the `avatar` folder, but the corresponding route folders do exist.

**Comment by annando** [comment]
My idea here is to create some console command to completely clear the avatar-cache-entries in the contact table. This would enforce a new caching.

**Comment by MrPetovan** [comment]
Ok, thank you for the information. I'm not sure what to do with it at the moment but it should help.

**Comment by csolisr** [comment]
I have no issues running commands on MySQL directly, in the meanwhile

**Comment by csolisr** [comment]
I built a small bash script that can be used to try resetting the missing avatar pictures, but I think I'm still missing something:

```
#Set your parameters here
url=yourfriendicasite.example.net
user=friendica
db=friendica
folder=/var/www/friendica
#Internal parameters:
#Amount of times the loop has run
iteration=0
#Number of invalid avatars found. Set to 1 initially so we can run the loop at least once
n=1
#Number of entries processed
nx=0
#Go to the Friendica installation
cd $folder
#Loop at least once, until no invalid avatars are found
until [[ $n -eq 0 ]]
do
        #Add to the loop, reset values
        iteration=$(expr $iteration + 1)
        n=0
        nx=0
        #Amount of items where the avatar is currently set to be stored in the local cache
        m=$(sudo mysql $db -B -N -q -e "select count(*) from contact where photo like 'https:\/\/$url/avatar/%'")
        #This loops through a SQL query listed below
        while read -r i j
        do
                nx=$(expr $nx + 1)
                folderescaped=${folder////\\/}
                #Substitute the URL path with the folder path so we can search for it in the local file system
                k=$(echo $j | sed -e "s/https:\/\/$url/$folderescaped/g" -e "s/\?ts=.*//g")
                if [[ ! -e $k ]]
                then
                        avatar=$(sudo mysql $db -B -N -q -e "select avatar from contact where id = \"$i\"")
                        #If we have a remote avatar as a fallback, download it
                        if [[ ! -z $avatar ]]
                        then
                                sudo -u $user curl $avatar -s -o $k && sudo chmod 755 $k && sudo chown $user:$user $k
                        else
                                sudo mysql $db -e "update contact set photo = \"\" where id = \"$i\""
                        fi
                        n=$(expr $n + 1)
                fi
                printf "\rIteration $iteration\tPhotos: $n\tEntry $nx/$m"
        done < <(sudo mysql $db -B -N -q -e "select id, photo from contact where photo like 'https:\/\/$url/avatar/%' order by id")
        printf "\nFixing folders and moving to avatar cache...\n"
        sudo find ./avatar -depth -not -user $user -exec sudo chown $user:$user {} \;
        sudo find ./avatar -depth -type f -not -perm 755 -exec sudo chmod 755 {} \;
        sudo -u $user bin/console movetoavatarcache &> /dev/null
done
```

**Comment by csolisr** [comment]
Quick follow-up to this ticket - I've been running the above code repeatedly for the last week, to no avail. The only viable way I have to restore the pictures is to manually open each of the affected profiles in my web browser so the picture is redownloaded. Is there a better way to automate this redownload?

**Comment by annando** [comment]
You could try to clear the `photo`, `thumb` and `micro` fields in the contact table. The system **should** then fetch the avatars the next time, this avatar is displayed. Please test it on a handful of contacts before you remove all entries. Also you should then clear the avatar folder completely to ensure that there aren't any orphaned entries.
